packages-ts/gauntlet-terra-contracts/src/lib/contracts.ts

```diff
@@ -28,109 +30,191 @@ export type TerraABI = {
+      throw new Error(`Read invalid contract address ${address} for ${this.id} contract from env`)
 
-export const getContractCode = async (contractId: CONTRACT_LIST, version): Promise<string> => {
-  if (version === 'local') {
+    logger.debug(`Using deployed instance of ${this.id}: ${name}=${address}`)
+    return this
+  }
+
+  loadContractCode = async (version: string): Promise<void> => {
+    version = version ? version : this.defaultVersion
+    assertions.assert(
+      !this.version || version == this.version,
+      `Loading multiple versions (${this.version} and ${version}) of the same contract is unsupported.`,
+    )
+    this.version = version
+
+    if (version === 'local') {
+      // Possible paths depending on how/where gauntlet is being executed
+      const possibleContractPaths = [
+        path.join(__dirname, '../../artifacts/bin'),
+        path.join(process.cwd(), './artifacts/bin'),
+        path.join(process.cwd(), './tests/e2e/common_artifacts'),
+        path.join(process.cwd(), './packages-ts/gauntlet-terra-contracts/artifacts/bin'),
+      ]
+
+      const codes = possibleContractPaths
+        .filter((contractPath) => existsSync(`${contractPath}/${this.id}.wasm`))
+        .map((contractPath) => {
+          const wasm = readFileSync(`${contractPath}/${this.id}.wasm`)
+          return wasm.toString('base64')
+        })
+      this.bytecode = codes[0]
+    } else {
+      const url = `${this.downloadUrl}${version}/${this.id}.wasm`
+      logger.loading(`Fetching ${url}...`)
+      const response = await fetch(url)
+      const body = await response.arrayBuffer()
+      if (body.length == 0) {
+        throw new Error(`Download ${this.id}.wasm failed`)
+      }
+      this.bytecode = Buffer.from(body).toString('base64')
+    }
+  }
+
+  loadContractABI = async (): Promise<void> => {
     // Possible paths depending on how/where gauntlet is being executed
+    const cwd = process.cwd()
     const possibleContractPaths = [
-      path.join(__dirname, '../../artifacts/bin'),
-      path.join(process.cwd(), './artifacts/bin'),
-      path.join(process.cwd(), './tests/e2e/common_artifacts'),
-      path.join(process.cwd(), './packages-ts/gauntlet-terra-contracts/artifacts/bin'),
+      path.join(__dirname, './artifacts/contracts'),
+      path.join(cwd, './contracts'),
+      path.join(cwd, '../../contracts'),
+      path.join(cwd, './packages-ts/gauntlet-terra-contracts/artifacts/contracts'),
     ]
 
-    const codes = possibleContractPaths
-      .filter((contractPath) => existsSync(`${contractPath}/${contractId}.wasm`))
+    const abi = possibleContractPaths
+      .filter((path) => existsSync(`${path}/${this.dirName}/schema`))
       .map((contractPath) => {
-        const wasm = readFileSync(`${contractPath}/${contractId}.wasm`)
-        return wasm.toString('base64')
+        const toPath = (type) => {
+          if (this.id == CONTRACT_LIST.CW20_BASE && type == 'execute_msg') {
```

Comment:
Why would the specific `cw20` contract get a different treatment than any other contract? This coupling needs to be removed and this fn rewritten to be more generalized.

Reply:
This was not new in this PR, just moved from here:
https://github.com/smartcontractkit/chainlink-terra/pull/204/files/#diff-fd94fe66fb4093bf76dd608bfb2120fb03b9e19cc7af090fcedf03a57ba68f13L105-L114

But makes sense to try and improve it here since we're refactoring so many things around it.

The schemas files in our repo were extracted from the tgz release artifacts in the `cw-plus` repo here:
https://github.com/CosmWasm/cw-plus/releases

All of the schema names follow the same pattern of `{query,instantiate,execute}_msg.json` except for cw20, which they decided to make a special case for some reason:
```
packages-ts/gauntlet-terra-contracts/artifacts/contracts//cw20_base/schema:
all_accounts_response.json   allowance_response.json      cw20_execute_msg.json        query_msg.json
all_allowances_response.json balance_response.json        instantiate_msg.json         token_info_response.json

packages-ts/gauntlet-terra-contracts/artifacts/contracts//cw3_flex_multisig/schema:
execute_msg.json     instantiate_msg.json query_msg.json

packages-ts/gauntlet-terra-contracts/artifacts/contracts//cw4_group/schema:
admin_response.json        instantiate_msg.json       member_response.json       total_weight_response.json
execute_msg.json           member_list_response.json  query_msg.json
```
So if we don't want it to be explicitly handled as a special case in our code, I see two possibilities forward:

1.  We rename `cw20_execute_msg.json` as `execute_msg.json` to match all other schemas.
or
3.  We create a new dictionary allowing each contract to have its own unique schema names, but have them all be identical except for this one.

I'm not sure which one is the best way forward.  1 seems like the simpler fix, but I'm hesitant to do that in case it might cause an unforeseen problems down the road (for example, title field--which we don't currently use--not matching filename).  What do you think?

Reply:
I think this is the only remaining issue, still waiting for feedback on which would be better for the long-term, (1) or (2)?  Or if this would require more research to be sure, maybe we just leave it for now?